kafka/src/main/java/org/apache/apex/malhar/kafka/KafkaSinglePortExactlyOnceOutputOperator.java

```diff
@@ -0,0 +1,369 @@
+    return false;
+  }
+
+  private boolean alreadyInKafka(T message)
+  {
+    if ( windowId <= windowDataManager.getLargestRecoveryWindow() ) {
+      return true;
+    }
+
+    if (partialWindowTuples.containsKey(message)) {
+
+      Integer val = partialWindowTuples.get(message);
+
+      if ( val == 0 ) {
+        return false;
+      } else if ( val == 1 ) {
+        partialWindowTuples.remove(message);
+      } else {
+        partialWindowTuples.put(message, val - 1);
+      }
+
+      return true;
+    }
+
+    return false;
+  }
+
+  private Map<Integer,Long> getPartitionsAndOffsets() throws ExecutionException, InterruptedException
+  {
+    List<PartitionInfo> partitionInfoList = consumer.partitionsFor(getTopic());
+    List<TopicPartition> topicPartitionList = new java.util.ArrayList<>();
+
+    for ( PartitionInfo partitionInfo: partitionInfoList) {
+      topicPartitionList.add(new TopicPartition(getTopic(), partitionInfo.partition()) );
+    }
+
+    Map<Integer,Long> parttionsAndOffset = new HashMap<>();
+    consumer.assign(topicPartitionList);
+
+    for ( PartitionInfo partitionInfo: partitionInfoList) {
+
+      try {
+
+        TopicPartition topicPartition = new TopicPartition(getTopic(), partitionInfo.partition());
+        consumer.seekToEnd(topicPartition);
+        parttionsAndOffset.put(partitionInfo.partition(), consumer.position(topicPartition));
+
+      } catch (Exception ex) {
+        throw new RuntimeException(ex);
+      }
+    }
+
+    return parttionsAndOffset;
+  }
+
+  private void rebuildPartialWindow()
+  {
+    logger.info("Rebuild the partial window after " + windowDataManager.getLargestRecoveryWindow());
+
+    Map<Integer,Long> storedOffsets = getStoredOffsets();
+    Map<Integer,Long> currentOffsets = getCurrentOffsets();
+
+    if (storedOffsets == null || currentOffsets == null) {
+      logger.debug("No tuples found while building partial window " + windowDataManager.getLargestRecoveryWindow());
+      return;
+    }
+
+    List<TopicPartition> topicPartitions = new ArrayList<>();
+
+    for (Map.Entry<Integer,Long> entry: currentOffsets.entrySet()) {
+
+      topicPartitions.add(new TopicPartition(getTopic(), entry.getKey()));
+    }
+
+    consumer.assign(topicPartitions);
+
+    for (Map.Entry<Integer,Long> entry: currentOffsets.entrySet()) {
+
+      Long storedOffset = 0L;
+      Integer currentPartition = entry.getKey();
+      Long currentOffset = entry.getValue();
+
+      if (storedOffsets.containsKey(currentPartition)) {
+        storedOffset = storedOffsets.get(currentPartition);
+      }
+
+      if (storedOffset >= currentOffset) {
+        continue;
+      }
+
+      consumer.seek(new TopicPartition(getTopic(), currentPartition), storedOffset);
+
+      int kafkaAttempt = 0;
+
+      while ( true ) {
+
+        ConsumerRecords<String, String> consumerRecords = consumer.poll(100);
+
+        if (consumerRecords.count() == 0) {
+          if (kafkaAttempt++ == KAFKA_CONNECT_ATTEMPT) {
```

Comment:
If some other operator are pushing data into same partition, it goes to infinite loop. The ConsumerRecord would give you an offset, you should break the while loop if the offset is bigger than the current offset you load before.